### PR TITLE
Networking broadcasting feature

### PR DESCRIPTION
**Pull request by frozenbanana**
- Broadcasting implemented and tested for UDP
- Namespace duplication cleaned
- Memory issues in packet reading fixed.

### CODE REVIEW

File: LambdaEngine/Include/Networking/API/ClientRemoteBase.h [hunk 1]
```diff
@@ -41,6 +41,7 @@ namespace LambdaEngine
 
 		void TransmitPackets();
 		void DecodeReceivedPackets();
+		bool SendReliableBroadcast(NetworkSegment* packet, IPacketListener* listener = nullptr);
```
Comment (TheRealChriAnd) [inline LambdaEngine/Include/Networking/API/ClientRemoteBase.h]: This function should be public

File: LambdaEngine/Include/Networking/API/NetworkSegment.h [hunk 2]
```diff
@@ -37,6 +37,7 @@ namespace LambdaEngine
 			TYPE_ACCEPTED				= UINT16_MAX - 7,
 			TYPE_NETWORK_ACK			= UINT16_MAX - 8,
 			TYPE_NETWORK_DISCOVERY		= UINT16_MAX - 9,
+			TYPE_BROADCAST				= UINT16_MAX - 10,
```
Comment (TheRealChriAnd) [inline LambdaEngine/Include/Networking/API/NetworkSegment.h]: There should not be a Segment Type named BROADCAST

File: LambdaEngine/Include/Networking/API/ServerBase.h [hunk 3]
```diff
@@ -50,6 +50,7 @@ namespace LambdaEngine
         uint8 GetClientCount();
         const ServerDesc& GetDescription() const;
         const ClientMap& GetClients() const;
+        bool SendReliableBroadcast(ClientRemoteBase* pClient, NetworkSegment* pPacket, IPacketListener* pListener);
```
Comment (TheRealChriAnd) [inline LambdaEngine/Include/Networking/API/ServerBase.h]: This function should be private

File: LambdaEngine/Source/Networking/API/ClientRemoteBase.cpp [hunk 4]
```diff
@@ -148,6 +148,19 @@ namespace LambdaEngine
...
+		if (!m_pServer->SendReliableBroadcast(this, packet, listener))
+		{
+			LOG_WARNING("[ClientRemoteBase]: Can not broadcast packet");
+			return false;
+		}
+
+		return true;
+	}
+
+
```
Comment (TheRealChriAnd) [inline LambdaEngine/Source/Networking/API/ClientRemoteBase.cpp]: Multiple newlines above and below the function definition. This should be a single newline.

packet should be pPacket
listener should be pListener

The if statement is unnecessary and can be omitted. Make the function return the result from SendReliableBroadcast directly

File: LambdaEngine/Source/Networking/API/ClientRemoteBase.cpp [hunk 5]
```diff
@@ -202,6 +215,7 @@ namespace LambdaEngine
 						}
 					}
 				}
+
```
Comment (TheRealChriAnd) [inline LambdaEngine/Source/Networking/API/ClientRemoteBase.cpp]: Remove the newline

File: LambdaEngine/Source/Networking/API/ServerBase.cpp [hunk 6]
```diff
@@ -187,7 +211,6 @@ namespace LambdaEngine
 
 		if (!m_ClientsToAdd.IsEmpty() || !m_ClientsToRemove.IsEmpty())
 		{
-			std::scoped_lock<SpinLock> lock2(m_LockClientVectors);
```
Comment (TheRealChriAnd) [inline LambdaEngine/Source/Networking/API/ServerBase.cpp]: This should not be moved

File: Server/Source/Server.cpp [hunk 7]
```diff
@@ -22,9 +22,16 @@
 
 #include "Math/Random.h"
 
+#include "Rendering/Renderer.h"
+
+
+#include "Networking/API/NetworkSegment.h"
+#include "Networking/API/BinaryEncoder.h"
```
Comment (TheRealChriAnd) [inline Server/Source/Server.cpp]: Duplicate. They are already included in PlatformNetworkUtils.h

File: Server/Source/Server.cpp [hunk 8]
```diff
@@ -75,22 +78,31 @@ bool Server::OnKeyPressed(const LambdaEngine::KeyPressedEvent& event)
...
+	}
+
+	// Simulate first Remote client broadcasting
+	if (m_pServer->GetClientCount() > 0)
+	{
+		auto chosenClientPair = m_pServer->GetClients().begin();
+		NetworkSegment* pPacket = chosenClientPair->second->GetFreePacket(99);
+		BinaryEncoder encoder(pPacket);
+		encoder.WriteString("Test broadcast from server.cpp");
+		m_pServer->SendReliableBroadcast(chosenClientPair->second, pPacket, nullptr);
```
Comment (TheRealChriAnd) [inline Server/Source/Server.cpp]: Should not be accessible from outside of ServerBase

**Review by TheRealChriAnd — CHANGES_REQUESTED** [review]
Please fix the things marked in the comments

**Comment by frozenbanana** [comment]
All review points have been resolved.

**Review by TheRealChriAnd — APPROVED** [review]
Looking great!

**Pull request merged**